File: TopCoder/src/main/java/com/leetcode/tree/BSTIterator.java
package com.leetcode.tree;

import java.util.Stack;

//https://leetcode.com/problems/binary-search-tree-iterator/
public class BSTIterator {
    private Stack<TreeNode> stack = new Stack<>();
    private TreeNode current;

    public BSTIterator(TreeNode root) {
        current = root;
    }

    /** @return the next smallest number */
    public int next() {
        int temp = 0;
        if (current != null || !stack.isEmpty()) {

            while (current!= null) {
                stack.push(current);
                current = current.left;
            }

            TreeNode peekNode = stack.pop();
            temp = peekNode.val;
            current = peekNode.right;
        }
        return temp;
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return current != null || !stack.isEmpty();
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(7);
        root.left = new TreeNode(3);

        root.right = new TreeNode(15);
        root.right.left = new TreeNode(9);
        root.right.right = new TreeNode(20);
        BSTIterator iterator = new BSTIterator(root);
        System.out.println(iterator.next());    // return 3
        System.out.println(iterator.next());    // return 7
        System.out.println(iterator.hasNext()); // return true
        System.out.println(iterator.next());    // return 9
        System.out.println(iterator.hasNext()); // return true
        System.out.println(iterator.next());    // return 15
        System.out.println(iterator.hasNext()); // return true
        System.out.println(iterator.next());    // return 20
        System.out.println(iterator.hasNext()); // return false
    }
}

File: TopCoder/python/BSTIterator.py
from typing import List

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class BSTIterator:

    def __init__(self, root: TreeNode):
        self.current: TreeNode = root
        self.stack: List[TreeNode] = []

    def next(self) -> int:
        """
        @return the next smallest number
        """
        next_val: TreeNode
        if self.current or self.stack:
            while self.current:
                self.stack.append(self.current)
                self.current = self.current.left
            next_val = self.stack.pop()
            self.current = next_val.right
        return next_val.val   


    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return self.current is not None or len(self.stack) > 0

if __name__ == "__main__":
    root = TreeNode(7)
    root.left = TreeNode(3)
    
    root.right = TreeNode(15)
    root.right.left = TreeNode(9)
    root.right.right = TreeNode(20)
    iterator = BSTIterator(root)
    print(iterator.next())
    print(iterator.next())
    print(iterator.hasNext())
    print(iterator.next())
    print(iterator.hasNext())
    print(iterator.next())
    print(iterator.hasNext())
    print(iterator.next())
    print(iterator.hasNext())